### PR TITLE
interpret: do not call machine read hooks during validation

### LINKED ISSUE #3347: Alloc access tracking shows spurious reads after each write
After each write we do validation which causes reads, and these reads actually trigger the `before_memory_read` machine hook and then cause alloc access tracking to print.

It seems to me like we probably want to not call the hook for such "fake" reads at all? These reads don't actually exist in the AM; usually the value would get validated "while copying" but Miri doesn't support that so we do some extra reads.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/miri/issues/3347

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 1]
```diff
@@ -380,16 +380,12 @@ pub fn eval_in_interpreter<'mir, 'tcx>(
         }
         Ok(mplace) => {
             // Since evaluation had no errors, validate the resulting constant.
-
-            // Temporarily allow access to the static_root_alloc_id for the purpose of validation.
-            let static_root_alloc_id = ecx.machine.static_root_alloc_id.take();
-            let validation = const_validate_mplace(&ecx, &mplace, cid);
-            ecx.machine.static_root_alloc_id = static_root_alloc_id;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:387]: I am a bit torn about whether `before_alloc_read` should be suppressed during validation. OTOH it is nice that we can now remove this code here. On the other hand, now if validation does do a read from this static, it will see `Uninit` and show a surprising error rather than triggering a cycle error. I don't know if it is even possible to do that, it would require a `copy_op_transmute` from the static to somewhere else.

If validation passes, we'll still get the cycle error from the actual read part of `copy_op`.

File: src/tools/miri/tests/pass/alloc-access-tracking.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#![feature(start)]
+#![no_std]
+//@compile-flags: -Zmiri-track-alloc-id=17 -Zmiri-track-alloc-accesses -Cpanic=abort
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/alloc-access-tracking.rs:3]: As an isolated no_std test, I think the alloc IDs here should be fully stable.

Comment (RalfJung) [reply src/tools/miri/tests/pass/alloc-access-tracking.rs:3]: Damn, somehow it is different on macOS... I guess I'll have to add some `only-target`.

**Comment by RalfJung** [comment]
> TLDR: the last commit implements a custom machine just for validation.

How's that supposed to work...?
It may for for const-eval which runs validation after interpretation, but in this PR I am primarily concerned with Miri. We need to be able to validate thing that are not in the global `tcx`.

**Comment by oli-obk** [comment]
No tests failed so I assumed I didn't change behaviour

But generally the scheme can be replicated in miri

File: src/tools/miri/tests/pass/alloc-access-tracking.stderr [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+   |         ^^^^^^^^^ write access to allocation with id 17
+   |
+   = note: BACKTRACE:
+   = note: inside `start` at $DIR/alloc-access-tracking.rs:LL:CC
+
+note: tracking was triggered
+  --> $DIR/alloc-access-tracking.rs:LL:CC
+   |
+LL |         *ptr = 42; // Crucially, only a write is printed here, no read!
+   |         ^^^^^^^^^ read access to allocation with id 17
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/alloc-access-tracking.stderr:23]: The fact that this line exists here shows that your patch doesn't achieve the goal of this PR.

Comment (oli-obk) [reply src/tools/miri/tests/pass/alloc-access-tracking.stderr:23]: Should probably turn that into an annotation 😅

Comment (RalfJung) [reply src/tools/miri/tests/pass/alloc-access-tracking.stderr:23]: I don't know how, this is a "pass" test.

**Comment by oli-obk** [comment]
@bors r+

I can explore my ideas in separate PRs
